designs/0031-deprecation-schedule.md

```diff
@@ -0,0 +1,254 @@
+   treatment of some things as valid identifiers and valid keywords at the same
+   time - syntax highlighting especially suffers:
+   ```stan
+       real<lower=0> x; // notice lower is highlighted
+       real<offset=0> y; // notice offset is not
+       int offset; // this is (partially) the reason why
+   ```
+
+# Prior art
+[prior-art]: #prior-art
+
+- The original inspiration for this schedule comes from the Python library
+  NumPy, with their deprecation policy laid out in
+  [NEP-23](https://numpy.org/neps/nep-0023-backwards-compatibility.html#implementing-deprecations-and-removals).
+  The NumPy project provides a similar example of a large open-source code base
+  whose minor version number has now exceeded 20 (meaning a large number of
+  changes since the last 'breaking'/major version).
+- Python itself defines a similar policy in
+  [PEP-387](https://www.python.org/dev/peps/pep-0387/) which requires warnings
+  emitted for 2 minor versions or 1 major version (e.g, a feature deprecated in
+  3.10 could be removed in 3.12 **or** 4.0)
+- The Java language clarified its [deprecation
+  policy](https://docs.oracle.com/javase/9/core/enhanced-deprecation1.htm#JSCOR-GUID-23B13A9E-2727-42DC-B03A-E374B3C4CE96)
+  in Java 9 (which is technically a minor version, 1.9, for the JDK). The main
+  motivation for clarification was the response of users being unsure if a
+  deprecated feature would ever actually be removed. There was a
+  [study](https://doi.org/10.1007/S10664-019-09713-W) of this change published
+  in 2019. It contains the following relevant passage and some interesting
+  discussion of user reactions to a deprecation warning.
+  > By not removing deprecated features from an API after a transition period
+  > has passed, API producers and the Java JDK developers themselves have
+  > cheapened the meaning of deprecation.
+- https://semver.org/ - the description of proper 'semantic versioning'. In the
+  strictest interpretation, Stan already deviates from this scheme with some
+  regularity, but it is a common understanding of the major.minor.patch
+  versioning.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Above I set the term of deprecations at 3 minor versions/1 year. This may be
+  too long for some features and too short for others, and is worth discussing.
+- We have many existing deprecations which predate this proposed policy. How
+  they should be handled or removed is not explicitly covered; the easiest path
+  is to consider the first version after this policy is enacted to be their
+  'deprecation version', and remove them 3 minor versions later.
+- This is independent of questions about larger features like the ODE solver
+  changes and the potential of a Stan version 3.0, but these should be
+  considered when thinking of the long term health of the project as well. How
+  to handle the 'major' or semantic changes to the language is not resolved by
+  this proposal.
```

Comment:
I think it is.  Anything that's not a minor deprecation as defined here will require a major version bump.  Is there anything else to say?

Reply:
You're right, I more mean that the question of "when will we do stan 3 and what will be in it" is orthogonal to this.